src/extra/components/ppu.py

```diff
@@ -0,0 +1,183 @@
+        self, data: DataCollection, ppu: Union[KeyData, SourceData, str] = None
+    ):
+        """
+
+        Args:
+            data (DataCollection):
+            ppu (Union[KeyData, SourceData, str], optional):
+                Specify a Pulse Picker Unit device to use, necessary if a run
+                contains more than one PPU. This can be any of:
+                  - The device name of the source.
+                  - A `SourceData` or [KeyData][extra_data.KeyData] of the
+                    control source (e.g. `HED_XTD6_PPU/MDL/PPU_TRIGGER`) of a
+                    PPU.
+                  - The alias name of either a `SourceData` or
+                    [KeyData][extra_data.KeyData] belonging to a PPU.
+                  - A unique (case-insensitive) substring of a PPU source name.
+
+        Raises:
+            KeyError: If we can't identify a unique PPU device from the
+            arguments.
+        """
+        self.data = data
+        self.device = _find_ppu(data, ppu)
+
+    @lru_cache()
+    def number_of_trains(self, train_id: int) -> int:
+        """Number of trains picked for the sequence starting at train_id.
+
+        Args:
+            train_id (int): train ID of the sequence start.
+        """
+
+        # The Dipole PPU-like device does not allow to pick multiple trains
+        if "trainTrigger.numberOfTrains" not in self.device.keys():
+            return 1
+        n_trains = self.device["trainTrigger.numberOfTrains"]
+        return int(n_trains.select_trains(by_id[[train_id]]).ndarray()[0])
+
+    def train_ids(
+        self, offset: int = 0, labelled: bool = False
+    ) -> Union[List[int], pd.Series]:
+        """All train IDs picked by the PPU.
+
+        Args:
+            offset (int, optional):
+                offset to add to the selected trains. Defaults to 0.
+            labelled (bool, optional):
+                Returns a Pandas Series if set to True, where this index represents the
+                trigger sequence a train ID is part of. Defaults to False.
+
+        Returns:
+            Union[List[int], pd.Series]: Train IDs picked by the PPU.
+        """
+        seq_start = self.device["trainTrigger.sequenceStart"].ndarray()
+        # The trains picked are the unique values of trainTrigger.sequenceStart
+        # minus the first (previous trigger before this run).
+        start_train_ids = np.unique(seq_start)[1:] + offset
+
+        train_ids = []
+        sequences = []
+        for seq, train_id in enumerate(start_train_ids):
+            span = self.number_of_trains(train_id)
+            train_ids.extend(np.arange(train_id, train_id + span).tolist())
+            sequences.extend([seq] * span)
+
+        log.info(
+            f"PPU device {self.device.source} triggered for {len(train_ids)} train(s) across {len(sequences)} sequence(s)."
+        )
+
+        if labelled:
+            train_ids = pd.Series(train_ids, index=sequences)
+        return train_ids
+
+    def trains(
```

Comment:
Would you consider a different name for this method? Both in EXtra-data as well as across existing components, we use `.trains()` to return an iterator over trains. This method however takes something that has train IDs and changes it.

Reply:
Good point. Do you have something in mind?  
- `filter`? `filter_trains`?
- `train_selection`? `select`? `select_trains`?
- `pick`? `pick_trains`?
- `take`? `take_trains`?

I'd probably go for `pick_trains`.

Reply:
`select` and `select_trains` also have different meanings already (though in the long run, I would love to pass this component to `DataCollection.select_trains`, i.e. have some kind of `TrainSelector` protocol)

From your list, I would probably also go for `pick_trains`, but it's unfortunate one way or the other that it's the "picking-thing" calling on the "thing-to-be-picked" 🤔 
Maybe @takluyver can comment on what sounds most natural.

Reply:
> I would love to pass this component to DataCollection.select_trains, i.e. have some kind of TrainSelector protocol

yes, me too!  
But do we have an idea of what/how we want to select trains or how this protocol can look like? I can easily think on how to do that specifically for the PPU device, but I can't imagine other cases for now (maybe you have more use cases as part of the data reduction).

Reply:
How about we invert the problem: Instead of having a method that selects trains on some object, have a property or method that returns the correspondig `extra_data.by_id` slice? Then selection becomes: `run.select_trains(ppu.train_sel)` (name obviously preliminary)

Whenever we decide on a universal `TrainSelector` protocol, we can simply add it.

Reply:
+1 to something like `ppu.train_sel`. For the split-sequence option, we could make PPU iterable, giving some sort of TrainSequence or TrainGroup object so you could do `[run.select_trains(seq.train_sel) for seq in ppu]`.